classes/local/store/object_file_system.php

```diff
@@ -737,7 +737,26 @@ protected function copy_from_external_to_local($contenthash) {
      */
     public function redirect_to_presigned_url($contenthash, $headers = array()) {
         try {
-            redirect($this->externalclient->generate_presigned_url($contenthash, $headers));
+            $signedurl = $this->externalclient->generate_presigned_url($contenthash, $headers);
+            if (headers_sent()) {
+                debugging('Headers already sent! This redirect may be incorrectly cached in browser');
+            } else {
+                // Remove all previously-set headers.
+                foreach (headers_list() as $header) {
+                    // Get header name (text before the colon).
+                    if (preg_match('~^([^:]+):~', $header, $matches)) {
+                        header_remove($matches[1]);
+                    }
+                }
+                // Set expires and cache-control values to match the presigned URL expiry, which may be
+                // different to values previously set.
+                header('Expires: '. gmdate('D, d M Y H:i:s', $signedurl->expiresat) .' GMT');
+                // Use 'private' to allow browser caching only, otherwise via a shared cache users might
+                // be able to redirect to content that was only supposed to be displayed to a different
+                // user.
+                header('Cache-Control: private, max-age=' . ($signedurl->expiresat - time()));
```

Comment:
Adding Cache-Control: private I *think* should be safe, but not optimal, in all cases, I've written up a new feature here but I am pretty sure it's not a blocker for this patch but would like some eyes to double check:

https://github.com/catalyst/moodle-tool_objectfs/issues/469

Reply:
Hmm interesting - so you are suggesting that we could look at the existing cache-control header value and if there is one which already includes 'public' then just use public instead of private here? I hadn't thought of that. I think we could do that here in this patch and then don't need the separate issue, it doesn't sound too risky (if there is a case where Moodle sets cache-control public for these URLs already when they shouldn't be public, then that would already be a bug in the same way, so hopefully it doesn't). I'll have a quick look at that and then comment again when done (writing this now just in case I get distracted before I can finish it).

Reply:
I've pushed a change to allow 'public' as a second commit on this branch (so we can squash it into the original commit if you like, or get rid of it if it's not a good idea). Obviously this doesn't do the full thing with actually using unsigned URLs that you were proposing, only makes the redirect caching public.

If I've understood it correctly, I don't think there is a security hole in leaking a URL that was signed for user A so that user B can see it, except that it allows user B to access that single file, which in this case is what we want to happen since Moodle was serving the file with cache-control public.

If we send it 'public' now then in previous behaviour before this change (ie before the code to remove all the headers) we would have been sending it with cache-control public already so it at least shouldn't worsen security...